src/util/query_exec_util.cpp

```diff
@@ -0,0 +1,283 @@
+#include "execution/compiler/executable_query.h"
+#include "execution/exec/execution_context.h"
+#include "execution/sql/ddl_executors.h"
+#include "execution/vm/vm_defs.h"
+#include "metrics/metrics_manager.h"
+#include "network/network_defs.h"
+#include "network/network_util.h"
+#include "network/postgres/statement.h"
+#include "optimizer/cost_model/abstract_cost_model.h"
+#include "optimizer/statistics/stats_storage.h"
+#include "parser/expression/constant_value_expression.h"
+#include "parser/postgresparser.h"
+#include "parser/variable_set_statement.h"
+#include "settings/settings_manager.h"
+#include "transaction/transaction_context.h"
+#include "transaction/transaction_manager.h"
+
+namespace noisepage::util {
+
+std::unique_ptr<util::QueryExecUtil> QueryExecUtil::ConstructThreadLocal(
+    common::ManagedPointer<util::QueryExecUtil> util) {
+  return std::make_unique<util::QueryExecUtil>(util->db_oid_, util->txn_manager_, util->catalog_, util->settings_,
+                                               util->stats_, util->optimizer_timeout_);
+}
+
+QueryExecUtil::QueryExecUtil(catalog::db_oid_t db_oid,
+                             common::ManagedPointer<transaction::TransactionManager> txn_manager,
+                             common::ManagedPointer<catalog::Catalog> catalog,
+                             common::ManagedPointer<settings::SettingsManager> settings,
+                             common::ManagedPointer<optimizer::StatsStorage> stats, uint64_t optimizer_timeout)
+    : db_oid_(db_oid),
+      txn_manager_(txn_manager),
+      catalog_(catalog),
+      settings_(settings),
+      stats_(stats),
+      optimizer_timeout_(optimizer_timeout),
+      cost_func_(nullptr) {
+  exec_settings_ = execution::exec::ExecutionSettings{};
+  exec_settings_.UpdateFromSettingsManager(settings_);
+}
+
+void QueryExecUtil::SetDefaultDatabase() {
+  auto *txn = txn_manager_->BeginTransaction();
+  SetDatabase(catalog_->GetDatabaseOid(common::ManagedPointer(txn), catalog::DEFAULT_DATABASE));
+  txn_manager_->Commit(txn, transaction::TransactionUtil::EmptyCallback, nullptr);
+}
+
+void QueryExecUtil::SetDatabase(catalog::db_oid_t db_oid) { db_oid_ = db_oid; }
+
+void QueryExecUtil::SetExecutionSettings(execution::exec::ExecutionSettings exec_settings) {
+  exec_settings_ = exec_settings;
+}
+
+void QueryExecUtil::ClearPlans() {
+  schemas_.clear();
+  exec_queries_.clear();
+}
+
+void QueryExecUtil::UseTransaction(common::ManagedPointer<transaction::TransactionContext> txn) {
+  NOISEPAGE_ASSERT(txn_ == NULL, "Nesting transactions not supported");
+  txn_ = txn.Get();
+  own_txn_ = false;
+}
+
+void QueryExecUtil::BeginTransaction() {
+  NOISEPAGE_ASSERT(txn_ == NULL, "Nesting transactions not supported");
+  txn_ = txn_manager_->BeginTransaction();
+  own_txn_ = true;
+}
+
+void QueryExecUtil::EndTransaction(bool commit) {
+  NOISEPAGE_ASSERT(txn_ != NULL, "Transaction has not started");
+  if (own_txn_) {
```

Comment:
wait, so it does nothing if the txn is owned by someone else? maybe this should be checked before call, and assert here that you own the txn?